src/Microsoft.TemplateEngine.Utils/SemanticVersion.cs

```diff
@@ -0,0 +1,674 @@
+        //    considered greater
+        private int CompareToPrereleaseInfo(string other)
+        {
+            if (string.Equals(PrereleaseInfo, other, StringComparison.OrdinalIgnoreCase))
+            {
+                return 0;
+            }
+
+            string[] thisPrerelease = PrereleaseInfo.Split('.');
+            string[] otherPrerelease = other.Split('.');
+
+            for (int i = 0; i < thisPrerelease.Length && i < otherPrerelease.Length; ++i)
+            {
+                int segmentCompare = CompareToPrereleaseInfoSegment(thisPrerelease[i], otherPrerelease[i]);
+
+                if (segmentCompare != 0)
+                {
+                    return segmentCompare;
+                }
+            }
+
+            return thisPrerelease.Length.CompareTo(otherPrerelease.Length);
+        }
+
+        private static int CompareToPrereleaseInfoSegment(string left, string right)
+        {
+            if (IsNumericSegment(left, out bool ignored))
+            {
+                if (IsNumericSegment(right, out ignored))
+                {
+                    int us = int.Parse(left, NumberStyles.None, CultureInfo.InvariantCulture);
+                    int them = int.Parse(right, NumberStyles.None, CultureInfo.InvariantCulture);
+                    int numericCompare = us.CompareTo(them);
+
+                    if (numericCompare != 0)
+                    {
+                        return numericCompare;
+                    }
+                }
+                else
+                {
+                    return -1;
+                }
+            }
+            else if (IsNumericSegment(right, out ignored))
+            {
+                return 1;
+            }
+
+            int segmentCompare = StringComparer.OrdinalIgnoreCase.Compare(left, right);
+
+            if (segmentCompare != 0)
+            {
+                return segmentCompare;
+            }
+
+            return 0;
+        }
+
+        private int CompareVersionInformation(SemanticVersion other)
+        {
+            int majorCompare = Major.CompareTo(other.Major);
+
+            if (majorCompare != 0)
+            {
+                return majorCompare;
+            }
+
+            int minorCompare = Minor.CompareTo(other.Minor);
+
+            if (minorCompare != 0)
+            {
+                return minorCompare;
+            }
+
+            int patchCompare = Patch.CompareTo(other.Patch);
+
+            if (patchCompare != 0)
+            {
+                return patchCompare;
+            }
+
+            return 0;
+        }
+
+        private int ComparePrereleaseInfo(SemanticVersion other)
+        {
+            if (PrereleaseInfo != null)
+            {
+                if (other.PrereleaseInfo != null)
+                {
+                    int prereleaseCompare = CompareToPrereleaseInfo(other.PrereleaseInfo);
+
+                    if (prereleaseCompare != 0)
+                    {
```

Comment:
Nit: This check isn't strictly necessary, could just return the prereleaeCompare value here. If it's 0, we'll fall through to the final return of 0. But it might be best to keep the check, in case other things change which cause the fallthrough to be different.